[MRG] Systematic confounds standardization

NicolasGensollen opened the pull request:
<!---
This is a suggested pull request template for nilearn.
It's designed to capture information we've found to be useful in reviewing pull requests.

If there is other information that would be helpful to include, please don't hesitate to add it!

See here for more information on what is expected for pull requests:
https://github.com/nilearn/nilearn/blob/master/CONTRIBUTING.rst#pull-requests
-->

<!-- Please indicate after the # which issue you're closing with this PR.
This is helpful for the maintainers AND will magically close the issue when this
pull request is merged!
If the PR closes multiple issues, includes "closes" before each one is listed.
https://help.github.com/articles/closing-issues-using-keywords -->
Adresses #2584 .

<!-- Please give a brief overview of what has changed in the PR.
If you're not sure what to write, consider it a note to the maintainers to indicate
what they should be looking for when they review the pull request. -->
Changes proposed in this pull request:

- Add a parameter `standardize_confounds` to `NiftiMasker` to allow the user to specify whether or not to standardize confounds, and what strategy to use (see #2584 for more details).
- Update broken tests 

With these modifications, the tests proposed by @pbellec in #2584 seem to pass (on my laptop at least...).

Todo:

- [x] Add specific tests (we could maybe add directly the tests of #2584 somewhere?)
- [x] Update other masker objects if needed

bthirion reviewed (commented):
Thanks @NicolasGensollen 
Let me suggest that you add a test that fails with previous behavior and now passes.
Best,

NicolasGensollen commented:
I believe this is ready for review.

Note: Not really sure about where to put the new tests. I can put them somewhere else if needed.

bthirion commented on nilearn/decomposition/base.py:
```diff
@@ -255,6 +255,9 @@ class BaseDecomposition(BaseEstimator, CacheMixin, TransformerMixin):
         If standardize is True, the time-series are centered and normed:
         their mean is put to 0 and their variance to 1 in the time dimension.
 
+    standardize_confounds: boolean, optional
```
Well, it is a Bool or a string...

bthirion commented on nilearn/decomposition/base.py:
```diff
@@ -323,8 +326,8 @@ class BaseDecomposition(BaseEstimator, CacheMixin, TransformerMixin):
                  mask=None, smoothing_fwhm=None,
-                 standardize=True, detrend=True,
-                 low_pass=None, high_pass=None, t_r=None,
+                 standardize=True, standardize_confounds=True,
```
We should probably have uniofrm defaults, e.g. True everywhere.

bthirion commented on nilearn/input_data/multi_nifti_masker.py:
```diff
@@ -49,6 +49,16 @@ class MultiNiftiMasker(NiftiMasker, CacheMixin):
         to zero mean and scaled to unit variance.
         False : Do not standardize the data.
 
+    standardize_confounds: {'zscore', 'psc', True, False}, default is 'zscore'
```
This seems to be the exact docstring for that bit of code.

bthirion commented on nilearn/input_data/multi_nifti_masker.py:
```diff
@@ -49,6 +49,16 @@ class MultiNiftiMasker(NiftiMasker, CacheMixin):
+        Strategy to standardize the confounds.
+        'zscore': the signal is z-scored. Timeseries are shifted
+        to zero mean and scaled to unit variance.
+        'psc':  Timeseries are shifted to zero mean value and scaled
```
I think that 'psc' does not make sense for confounds. I'd rather not have it.

bthirion commented on nilearn/tests/test_masking.py:
```diff
@@ -30,6 +32,80 @@
+
+    # Z-score signal and default for the confounds
+    assert(_confounds_regression(standardize_signal=True) < 10. * eps)
+
```
But is one of these tests failing with the previous verison of the code ?

bthirion reviewed (commented):
Thx ! I left a couple of comments.

NicolasGensollen commented:
Thanks for the comments @bthirion. 

Since `pcs` does not make much sense for `standardize_confounds`, I think that there is no need for this parameter to behave like `standardize`, and having it as a simple boolean should be enough. I changed it to be an optional boolean parameter defaulting to `True`, meaning that confounds are z-scored by default, as suggested in #2584.

In addition, it seems like the default values for `standardize` are not always consistent between docstrings and code (see #2612). We might want to address this in a separate PR.

For the tests, since I changed the API, they won't run with previous versions of Nilearn as is. Previous to this PR, `standardize` was used for both signals and confounds standardization (which can now be reproduced by setting `standardize_confounds` according to `standardize`). 

The test code from #2584 with previous versions of nilearn displays 2 failures:

- one corresponding to the case where `standardize=False` and no demeaning of confounds 
- one corresponding to the case where `standardize='psc'` and demaning of confounds

My understanding is that, one of the reasons the first test fails, is because `signal.clean` still divides confounds by the max value when standardize is `False` as can be seen here:

https://github.com/nilearn/nilearn/blob/82f4075d8a8ea9aec25e66bd87ebb79a6be6d32f/nilearn/signal.py#L612-L621

while the second test fails because confounds are standardized with `psc` (since the same parameter is used for signal and confounds) while the code from #2584 does a z-score instead.

If you run the same tests with the fix of this PR, they should all pass.

Let me know if this is not clear and if we should have more tests for this.

bthirion reviewed (approved):
LGTM, thx !

bthirion commented:
@pbellec Any feedback on this ?

NicolasGensollen commented:
Does someone else want to take a look, or should consider this as ready for merging? Thx!

bthirion commented:
Merging, thx !
